src/variety/myopia.js

```diff
@@ -0,0 +1,439 @@
+				if (cell0.qnum <= 0) {
+					continue;
+				}
+				var row = [cell0, -1, -1, -1, -1];
+				for (var dir = 1; dir <= 4; dir++) {
+					var addr = cell0.getaddr();
+					addr.movedir(dir, 1);
+					while (!addr.getb().isnull && !addr.getb().isLine()) {
+						addr.movedir(dir, 2);
+					}
+					if (addr.getb().isLine()) {
+						row[dir] =
+							(Math.abs(dir >= 3 ? addr.bx - cell0.bx : addr.by - cell0.by) +
+								1) /
+							2;
+					}
+				}
+				console.log(row);
+				ret.push(row);
+			}
+			return (this._info.lineDirs = ret);
+		},
+
+		checkLineDirExist: function() {
+			var clues = this.getLineDirs();
+			for (var i in clues) {
+				for (var dir = 1; dir <= 4; dir++) {
+					if (!(clues[i][0].qnum & (1 << (dir - 1)))) {
+						continue;
+					}
+					if (clues[i][dir] === -1) {
+						this.failcode.add("arNoLineSeen");
+						if (this.checkOnly) {
+							return;
+						}
+						clues[i][0].seterr(4 << dir);
+					}
+				}
+			}
+		},
+
+		checkLineDirCloser: function() {
+			var clues = this.getLineDirs();
+			var unknown = this.board.cols + this.board.rows;
+			for (var i in clues) {
+				var mindist = unknown;
+				for (var dir = 1; dir <= 4; dir++) {
+					if (clues[i][dir] === -1) {
+						continue;
+					}
+					if (clues[i][0].qnum & (1 << (dir - 1))) {
+						mindist = Math.min(mindist, clues[i][dir]);
+					}
+				}
+				for (var dir = 1; dir <= 4; dir++) {
+					var dist = clues[i][dir];
+					if (clues[i][0].qnum & (1 << (dir - 1))) {
+						continue;
+					}
+					if (mindist === unknown && dist > 1) {
+						continue;
+					}
+					if (dist !== -1 && dist <= mindist) {
+						this.failcode.add("arDistanceGt");
+						if (this.checkOnly) {
+							return;
+						}
+						clues[i][0].seterr(0x7c);
+
+						var addr = clues[i][0].getaddr();
+						for (var n = 0; n < dist - 1; n++) {
+							addr.movedir(dir, 2);
+							addr.getc().seterr(1);
+						}
+						addr.movedir(dir, 1);
+						addr.getb().seterr(1);
+					}
+				}
+			}
+		},
+
+		checkLineDirUnequal: function() {
+			var clues = this.getLineDirs();
+			var unknown = this.board.cols + this.board.rows;
+			for (var i in clues) {
+				var mindist = unknown;
+				for (var dir = 1; dir <= 4; dir++) {
+					if (clues[i][dir] === -1) {
+						continue;
+					}
+					if (
+						clues[i][0].qnum & (1 << (dir - 1)) &&
+						mindist !== clues[i][dir]
+					) {
+						mindist = mindist === unknown ? clues[i][dir] : -1;
+					}
+				}
+				for (var dir = 1; dir <= 4; dir++) {
+					var dist = clues[i][dir];
+					if (!(clues[i][0].qnum & (1 << (dir - 1)))) {
+						continue;
+					}
+					if (dist !== -1 && dist !== mindist) {
+						this.failcode.add("arDistanceNe");
+						if (this.checkOnly) {
+							return;
+						}
+						clues[i][0].seterr(4 << dir);
+
+						var addr = clues[i][0].getaddr();
+						for (var n = 0; n < dist - 1; n++) {
+							addr.movedir(dir, 2);
+							addr.getc().seterr(1);
+						}
+						addr.movedir(dir, 1);
+						addr.getb().seterr(1);
```

Comment:
When marking a loop section as red, I prefer if the rest of the loop is gray.
`this.board.border.setnoerr();`
(applies to the other function as well)